### PR TITLE
Merge/nex 778/task log

### PR DESCRIPTION
**Pull request by bartlomiejmarszal**
Apply changes in TaskLog to work with both NewSql and Rds databases

https://oat-sa.atlassian.net/browse/NEX-778

### CODE REVIEW

File: models/classes/notification/Notification.php [hunk 1]
```diff
@@ -53,71 +58,61 @@ class Notification implements JsonSerializable
...
-     * @param string $message
-     * @param string $senderId
-     * @param string $senderName
-     * @param string|null $id
-     * @param string|null $createdAt
-     * @param string|null $updatedAt
-     * @param int $status
      */
-    public function __construct($userId, $title, $message, $senderId, $senderName, $id = null, $createdAt = null, $updatedAt = null, $status = 0)
+    public function __construct(string $userId, string $title, string $message, string $senderId, string $senderName, ?string $id = null, ?string $createdAt = null, ?string $updatedAt = null, ?int $status = 0)
```
Comment (gabrielfs7) [inline models/classes/notification/Notification.php:62]:
```suggestion
    public function __construct(string $userId, string $title, string $message, string $senderId, string $senderName, string $id = null, string $createdAt = null, string $updatedAt = null, int $status = 0)
```

A default value is already being provided

File: models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php [hunk 2]
```diff
@@ -15,102 +15,68 @@
...
-    private $containerName;
-
     /**
      * RdsTaskLogBroker constructor.
      *
      * @param string $persistenceId
      * @param null $containerName
      */
-    public function __construct($persistenceId, $containerName = null)
+    public function __construct(string $persistenceId, ?string $containerName = null)
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php:64]:
```suggestion
    public function __construct(string $persistenceId, string $containerName = null)
```
A default value is already being provided

File: models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php [hunk 3]
```diff
@@ -146,40 +112,51 @@ public function createContainer()
     /**
      * @inheritdoc
      */
-    public function add(TaskInterface $task, $status, $label = null)
+    public function add(TaskInterface $task, string $status, ?string $label = null): void
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php:115]:
```suggestion
    public function add(TaskInterface $task, string $status, string $label = null): void
```
Default value already provided

File: models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php [hunk 4]
```diff
@@ -146,40 +112,51 @@ public function createContainer()
...
+            . ', '
+            . \common_Utils::toHumanReadablePhpString($this->containerName)
+            . ')';
     }
 
     /**
      * @inheritdoc
      */
-    public function updateStatus($taskId, $newStatus, $prevStatus = null)
+    public function updateStatus(string $taskId, string $newStatus, ?string $prevStatus = null): int
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php:159]:
```suggestion
    public function updateStatus(string $taskId, string $newStatus, string $prevStatus = null): int
```

Default value already provided

File: models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php [hunk 5]
```diff
@@ -201,7 +178,7 @@ public function updateStatus($taskId, $newStatus, $prevStatus = null)
     /**
      * @inheritdoc
      */
-    public function addReport($taskId, Report $report, $newStatus = null)
+    public function addReport(string $taskId, Report $report, ?string $newStatus = null): int
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/RdsTaskLogBroker.php:181]:
```suggestion
    public function addReport(string $taskId, Report $report, string $newStatus = null): int
```

Default value already provided

File: models/classes/taskQueue/TaskLog/Broker/TaskLogBrokerInterface.php [hunk 6]
```diff
@@ -34,138 +34,89 @@
...
     /**
      * Inserts a new task log with status for a task.
-     *
-     * @param TaskInterface $task
-     * @param string        $status
-     * @param null|string   $label
-     * @return void
      */
-    public function add(TaskInterface $task, $status, $label = null);
+    public function add(TaskInterface $task, string $status, ?string $label = null): void;
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/TaskLogBrokerInterface.php:64]:
```suggestion
    public function add(TaskInterface $task, string $status, string $label = null): void;
```
Default value already provided

File: models/classes/taskQueue/TaskLog/Broker/TaskLogBrokerInterface.php [hunk 7]
```diff
@@ -34,138 +34,89 @@
...
      *
      * The previous status can be used for querying the record.
      *
-     * @param string      $taskId
-     * @param string      $newStatus
-     * @param string|null $prevStatus
      * @return int count of touched records
      */
-    public function updateStatus($taskId, $newStatus, $prevStatus = null);
+    public function updateStatus(string $taskId, string $newStatus, ?string $prevStatus = null): int;
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/TaskLogBrokerInterface.php:73]:
```suggestion
    public function updateStatus(string $taskId, string $newStatus, string $prevStatus = null): int;
```
Default value already provided

File: models/classes/taskQueue/TaskLog/Broker/TaskLogBrokerInterface.php [hunk 8]
```diff
@@ -34,138 +34,89 @@
...
     /**
      * Add a report for a task. New status can be supplied as well.
-     *
-     * @param string      $taskId
-     * @param Report      $report
-     * @param null|string $newStatus
-     * @return int
      */
-    public function addReport($taskId, Report $report, $newStatus = null);
+    public function addReport(string $taskId, Report $report, ?string $newStatus = null): int;
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/TaskLogBrokerInterface.php:83]:
```suggestion
    public function addReport(string $taskId, Report $report, string $newStatus = null): int;
```
Default value already provided

File: test/unit/model/taskQueue/TaskLog/Decorator/RedirectUrlEntityDecoratorTest.php [hunk 9]
```diff
@@ -131,7 +131,7 @@ protected function getFixtureEntity($status = TaskLogInterface::STATUS_COMPLETED
                 'children' => []
             ],
             'master_status' => true
-        ]);
+        ], \DateTime::RFC3339);
```
Comment (gabrielfs7) [inline test/unit/model/taskQueue/TaskLog/Decorator/RedirectUrlEntityDecoratorTest.php:134]: Import namespace

File: test/unit/model/taskQueue/TaskLog/Entity/TaskLogEntityTest.php [hunk 10]
```diff
@@ -51,13 +51,11 @@ public function testEntityCreated()
                 'children' => []
             ],
             'master_status' => true
-        ]);
+        ], \DateTime::RFC3339);
```
Comment (gabrielfs7) [inline test/unit/model/taskQueue/TaskLog/Entity/TaskLogEntityTest.php:54]: import namespace

File: test/unit/model/taskQueue/TaskLog/Entity/TaskLogEntityTest.php [hunk 11]
```diff
@@ -97,7 +91,7 @@ public function testCreateWithReportNull()
             'created_at' => '2017-02-01 12:00:01',
             'updated_at' => '2017-02-01 14:00:01',
             'report' => [],
-        ]);
+        ], \DateTime::RFC3339);
```
Comment (gabrielfs7) [inline test/unit/model/taskQueue/TaskLog/Entity/TaskLogEntityTest.php:94]: import namespace

File: test/unit/model/taskQueue/TaskLog/TaskLogCollectionTest.php [hunk 12]
```diff
@@ -108,6 +108,6 @@ protected function createCollection()
                    'children' => []
                ],
             ],
-        ]);
+        ], \DateTime::RFC3339);
```
Comment (gabrielfs7) [inline test/unit/model/taskQueue/TaskLog/TaskLogCollectionTest.php:111]: import namespace

File: models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+    /**
+     * @inheritdoc
+     */
+    abstract public function getTableName(): string;
+
+    /**
+     * @inheritdoc
+     */
+    abstract public function add(TaskInterface $task, string $status, ?string $label = null): void;
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php:74]:
```suggestion
    abstract public function add(TaskInterface $task, string $status, string $label = null): void;
```

File: models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+    /**
+     * @inheritdoc
+     */
+    abstract public function updateStatus(string $taskId, string $newStatus, ?string $prevStatus = null): int;
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php:79]:
```suggestion
    abstract public function updateStatus(string $taskId, string $newStatus, string $prevStatus = null): int;
```

File: models/classes/taskQueue/TaskLog/Entity/TaskLogEntity.php [hunk 15]
```diff
@@ -118,13 +118,27 @@ public static function createFromArray(array $row)
...
-            isset($row[TaskLogBrokerInterface::COLUMN_CREATED_AT]) ? DateTime::createFromFormat('Y-m-d H:i:s', $row[TaskLogBrokerInterface::COLUMN_CREATED_AT], new \DateTimeZone('UTC')) : null,
-            isset($row[TaskLogBrokerInterface::COLUMN_UPDATED_AT]) ? DateTime::createFromFormat('Y-m-d H:i:s', $row[TaskLogBrokerInterface::COLUMN_UPDATED_AT], new \DateTimeZone('UTC')) : null,
+            self::parseDateTime($row, TaskLogBrokerInterface::COLUMN_CREATED_AT, $dateFormat),
+            self::parseDateTime($row, TaskLogBrokerInterface::COLUMN_UPDATED_AT, $dateFormat),
             Report::jsonUnserialize($row[TaskLogBrokerInterface::COLUMN_REPORT]),
             isset($row[TaskLogBrokerInterface::COLUMN_MASTER_STATUS]) ? $row[TaskLogBrokerInterface::COLUMN_MASTER_STATUS] : false
         );
     }
 
+    public static function parseDateTime(array $row, string $key, string $dateFormat): ?DateTime
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Entity/TaskLogEntity.php:128]: If this method is must be public and used by another classes, probably it should not be part of this class.

- [ ] Make method private

File: scripts/install/InstallNotificationTable.php [hunk 16]
```diff
@@ -21,64 +21,36 @@
...
-            ]
-        );
-
-        $this->getServiceManager()->register(NotificationServiceInterface::SERVICE_ID, $queue);
+    /**
+     * @return PersistenceManager
+     */
+    private function getPersistenceManager()
+    {
+        return $this->getServiceLocator()->get(PersistenceManager::SERVICE_ID);
```
Comment (gabrielfs7) [inline scripts/install/InstallNotificationTable.php:54]:
```suggestion
    private function getPersistenceManager(): PersistenceManager
    {
        return $this->getServiceLocator()->get(PersistenceManager::SERVICE_ID);
```

**Comment by mike-ionut-mihai-sandu-tao** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function count(TaskLogFilter $filter): int
+    {
+        try {
+            $qb = $this->getCountQuery($filter);
+            return (int)$qb->execute()->fetchColumn();
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php:141]:
```suggestion
            return (int)$this->getCountQuery($filter)
                ->execute()
                ->fetchColumn();
```

File: models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    private function getSearchQuery(TaskLogFilter $filter): QueryBuilder
+    {
+        $qb = $this->getQueryBuilder()
+            ->select($filter->getColumns())
+            ->from($this->getTableName());
+
+        $qb->setMaxResults($filter->getLimit());
+        $qb->setFirstResult($filter->getOffset());
```
Comment (gabrielfs7) [inline models/classes/taskQueue/TaskLog/Broker/AbstractTaskLogBroker.php:156]:
```suggestion
        $qb = $this->getQueryBuilder()
            ->select($filter->getColumns())
            ->from($this->getTableName())
            ->setMaxResults($filter->getLimit())
            ->setFirstResult($filter->getOffset());
```
We can get leverage from the fluent interface unless we cannot use it.

**Pull request merged**